cap/modules/mail/attributes.py

```diff
@@ -0,0 +1,258 @@
+      to be retrieved by the db
+    - conditions: a dict that defines a path, and what should be true for
+      that path, according to the record data, e.g.
+      {"path": "foo.bar.options",
+       "if": ["exists", "is_not_in"],
+       "values": [true, "No"],
+       "op": "and",
+       "mail": ["test@cern.ch"]}
+    All the used condition methods (found in the 'if' field),
+    are in the conditions.py file.
+    """
+    recipients = []
+    cc = []
+    bcc = []
+
+    owner = config.get('owner', {})
+    current_user = config.get('current_user', {})
+
+    if owner:
+        _owner = CONDITION_METHODS['owner'](record)
+        if owner.get('recipients'):
+            recipients.append(_owner)
+        if owner.get('cc'):
+            cc.append(_owner)
+        if owner.get('bcc'):
+            bcc.append(_owner)
+
+    if current_user:
+        _current = CONDITION_METHODS['current_user'](record)
+        if current_user.get('recipients'):
+            recipients.append(_current)
+        if current_user.get('cc'):
+            cc.append(_current)
+        if current_user.get('bcc'):
+            bcc.append(_current)
+
+    # TODO: update for experiment, roles (read/admin)
+
+    # apply rules for conditions
+    conditions = config.get('conditions', [])
+    for condition in conditions:
+        if check_condition(record, condition):
+            update_mail_lists_from_defaults(record, condition, recipients, cc, bcc)  # noqa
+
+    # remove duplicates
+    return list(set(recipients)), list(set(cc)), list(set(bcc))
+
+
+def generate_recipients(record, config):
+    """
+    Recipients generator for notification action.
+    Retrieves from config and from custom function if it exists:
+    - custom function name in the 'func' field of the recipients config
+    - implementation should ALWAYS go in the mail.custom.recipients.py file
+
+    Also returns the `type` of the recipients that the mail should have, in
+    this case: recipients/bcc/cc.
+    """
+    re_config = config.get('recipients')
+    if not re_config:
+        return
+
+    recipients, cc, bcc = get_recipients_from_config(record, re_config)
+    update_mail_lists_from_defaults(record, re_config, recipients, cc, bcc)
+
+    # get the mails from functions (adds to right list depending on config)
+    func = re_config.get('func')
+    if func:
+        if func.get('recipients'):
+            custom_recipients_func = getattr(custom_recipients, func['recipients'])  # noqa
+            recipients += custom_recipients_func(record, re_config)
+        if func.get('cc'):
+            custom_recipients_func = getattr(custom_recipients, func['cc'])
+            cc += custom_recipients_func(record, re_config)
+        if func.get('bcc'):
+            custom_recipients_func = getattr(custom_recipients, func['bcc'])
+            bcc += custom_recipients_func(record, re_config)
+
+    return recipients, cc, bcc
+
+
+def generate_message(record, config, action):
```

Comment:
- Check L177, if there is no default it will break
- L186 better handling

Reply:
For both message and subject, the defaults are being handled inside the `populate_template_from_ctx` function. maybe this should change, we can talk about it.